runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/environment/DockerContainerEnvironment.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.fnexecution.environment;
+
+import org.apache.beam.model.pipeline.v1.RunnerApi.Environment;
+import org.apache.beam.runners.fnexecution.control.InstructionRequestHandler;
+import org.apache.beam.runners.fnexecution.control.SdkHarnessClient;
+
+/**
+ * A {@link RemoteEnvironment} that wraps a running Docker container.
+ *
+ * <p>A {@link DockerContainerEnvironment} owns both the underlying docker container that it
+ * communicates with an the {@link InstructionRequestHandler} that it uses to do so.
+ *
+ * <p>Accessors are thread-compatible.
```

Comment:
Maybe worth referencing what this means.

From my read of Effective Java point 70, this would just be "not thread-safe" in that book's breakdown.

Reply:
"Thread-compatible" in this case means that methods that do not mutate the object do not require external synchronization. From my read that means "conditionally thread-safe", but it's only worth mentioning if `InstructionRequestHandler` itself is thread-safe and we expect clients to share these (which I believe to be true).

I'm not sure how to word this though. Do I just say "conditionally thread-safe" and assume it's understood that the thread-safe methods are the non-mutating ones?  This is implied by "thread-compatible", but I'm not sure how widely used that is.

Reply:
That depends. `close()` implementations usually block until completion. For example, we (will) depend on this behavior when executing ExecutableStages on runners in order to await completion and surface exceptions during bundle processing. Using an `AtomicBoolean` this way means that some calls may return before the container has been killed. Whether that's considered "safe" depends on how it's used, but I'm hesitant to go that far.

Reply:
Hold on, then. If close isn't thread-safe but should block until completion, doesn't it have to just be synchronized regardless? Especially because I wouldn't really consider that to be a performance-critical piece of code.

If so please do

Reply:
Done.